AX: Add AXTextMarker and AXTextMarkerRange DebugDescription APIs for client debugging.

AndresGonzalezApple opened the pull request:
#### 5deaf7c534b2895e6c3445fcfeb05fda6426c394
<pre>
AX: Add AXTextMarker and AXTextMarkerRange DebugDescription APIs for client debugging.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290383">https://bugs.webkit.org/show_bug.cgi?id=290383</a>
&lt;<a href="https://rdar.apple.com/problem/147836473">rdar://problem/147836473</a>&gt;

Reviewed by Tyler Wilcock.

For accessibility clients, AXTextMarkers and AXTextMarkerRanges are opaque pointers and there is no way of gaining any visibility into their values. This change adds the ability for a client to get a debug description of the TextMarkers to aid in debugging.

* LayoutTests/accessibility/text-marker/text-marker-debug-description-expected.txt: Added.
* LayoutTests/accessibility/text-marker/text-marker-debug-description.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-sonoma/TestExpectations:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::debugDescription const):
(WebCore::AXTextMarkerRange::debugDescription const):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper showNodeTreeForTextMarker:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
(-[WebAccessibilityObjectWrapper debugDescriptionForTextMarker:]): Deleted.
(-[WebAccessibilityObjectWrapper debugDescriptionForTextMarkerRange:]): Deleted.
(formatForDebugger): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKAccessibilityRootObject):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.cpp:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::textMarkerDebugDescription):
(WTR::AccessibilityUIElement::textMarkerRangeDebugDescription):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/cocoa/AccessibilityTextMarkerRangeCocoa.mm:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::textMarkerDebugDescription):
(WTR::AccessibilityUIElement::textMarkerRangeDebugDescription):

Canonical link: <a href="https://commits.webkit.org/292698@main">https://commits.webkit.org/292698@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0bf81001d06a8dc42eb3733339fa011425d17fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16407 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6566 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101866 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98838 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73757 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30969 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99796 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87538 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54092 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12333 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5351 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46641 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5435 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103889 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23861 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17404 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82806 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24112 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83607 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82195 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20656 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26851 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4388 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17339 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23823 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28978 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23482 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26962 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25223 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3556,6 +3510,24 @@ - (id)accessibilityAttributeValue:(NSString*)attribute forParameter:(id)paramete
+        [self showNodeTreeForTextMarker:textMarker];
+        return nil;
+    }
+#endif
```
For larger compiler-pragma blocks, I've found it nice to comment `#endif`s with the condition they are closing, i.e.:

`#endif // ENABLE(TREE_DEBUGGING)`

AndresGonzalezApple replied:
Fixed.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -225,10 +225,22 @@ String AXTextMarker::debugDescription() const
+#endif
+
+    return makeString(ids
+        , "role "_s, object ? accessibilityRoleToString(object->roleValue()) : "no object"_str
```
Is `_str` different from `_s`? If not, I think it could be better to be consistent with everything else that uses `_s`.

AndresGonzalezApple replied:
Fixed.

AndresGonzalezApple commented on LayoutTests/accessibility/text-marker/text-marker-debug-description-expected.txt:
```diff
@@ -0,0 +1,34 @@
+
+button:
+text: '￼OK'
+start: {role Button, anchor 0, affinity 1, offset 1, characterStart 0, characterOffset 0, origin Unknown}
```
+text: 'OK'

AndresGonzalezApple committed:
AX: Add AXTextMarker and AXTextMarkerRange DebugDescription APIs for client debugging.
https://bugs.webkit.org/show_bug.cgi?id=290383
<rdar://problem/147836473>

Reviewed by Tyler Wilcock.

For accessibility clients, AXTextMarkers and AXTextMarkerRanges are opaque pointers and there is no way of gaining any visibility into their values. This change adds the ability for a client to get a debug description of the TextMarkers to aid in debugging.

* LayoutTests/accessibility/text-marker/text-marker-debug-description-expected.txt: Added.
* LayoutTests/accessibility/text-marker/text-marker-debug-description.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-sonoma/TestExpectations:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::debugDescription const):
(WebCore::AXTextMarkerRange::debugDescription const):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper showNodeTreeForTextMarker:]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
(-[WebAccessibilityObjectWrapper debugDescriptionForTextMarker:]): Deleted.
(-[WebAccessibilityObjectWrapper debugDescriptionForTextMarkerRange:]): Deleted.
(formatForDebugger): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(WKAccessibilityRootObject):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.cpp:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarker.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.cpp:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityTextMarkerRange.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::textMarkerDebugDescription):
(WTR::AccessibilityUIElement::textMarkerRangeDebugDescription):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/cocoa/AccessibilityTextMarkerRangeCocoa.mm:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::AccessibilityUIElement::textMarkerDebugDescription):
(WTR::AccessibilityUIElement::textMarkerRangeDebugDescription):

Canonical link: https://commits.webkit.org/292698@main
